Fire onChanged events for Web Extension Storage API.

kiaraarose opened the pull request:
#### 62c9865813d8e9d50461961249155ac3de478da3
<pre>
Fire onChanged events for Web Extension Storage API.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267965">https://bugs.webkit.org/show_bug.cgi?id=267965</a>
<a href="https://rdar.apple.com/121474442">rdar://121474442</a>

Reviewed by Timothy Hatcher.

This patch fires storage.onChanged and storageArea.onChanged events when new data is added or
old data is removed from storage. The onChanged data sent to the listener only contains data
for key value pairs where the value for that key has changed. If no data has changed, don&apos;t
fire the event.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::dictionaryWithKeys):
Helper method that returns a new dictionary based on the one provided, containing the specified keys.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
(WebKit::WebExtensionContext::storageSet):
(WebKit::WebExtensionContext::storageRemove):
(WebKit::WebExtensionContext::storageClear):
(WebKit::WebExtensionContext::fireStorageChangedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm:
(WebKit::WebExtensionAPIStorage::storageAreaForType):
(WebKit::WebExtensionContextProxy::dispatchStorageChangedEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/273942@main">https://commits.webkit.org/273942@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1d1b442a49cde2ec630589c33b230c691cc5cd5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/37304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/16197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/39584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/39850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/33277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/38610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/18789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/13345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/39850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/37869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/18789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/39584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/39850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/18789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/39584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/41110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/18789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/39584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/41110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/12238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/13345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/41110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/39584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4824 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/Platform/cocoa/CocoaHelpers.h:
```diff
@@ -108,6 +108,7 @@ NSString *encodeJSONString(id, JSONOptionSet = { }, NSError ** = nullptr);
 NSData *encodeJSONData(id, JSONOptionSet = { }, NSError ** = nullptr);
 
 NSDictionary *dictionaryWithLowercaseKeys(NSDictionary *);
+NSDictionary *keepObjectsWithKeys(NSDictionary *, NSArray *keys);
```
open to a better name for this

jeffmapple replied:
`dictionaryContainingObjectsWithKeys()`?

xeenon replied:
`dictionaryWithKeys()`?

kiaraarose replied:
both are great options! leaning towards `dictionaryWithKeys`

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -147,13 +141,20 @@
     auto storage = storageForType(storageType);
-    [storage deleteValuesForKeys:createNSArray(keys).get() completionHandler:makeBlockPtr([&, completionHandler = WTFMove(completionHandler)](NSString *errorMessage) mutable {
-        if (errorMessage)
+    [storage getValuesForKeys:createNSArray(keys).get() completionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, keys, storageType, storage, completionHandler = WTFMove(completionHandler)](NSDictionary<NSString *, NSString *> *oldValuesAndKeys, NSString *errorMessage) mutable {
```
should this also use `retainPtr`? `storage = RetainPtr { storage }`

xeenon replied:
Yes, or get the storage inside the block again with `storageForType(storageType)`.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -191,6 +199,48 @@
+    if (!oldKeysAndValues.count && !newKeysAndValues.count)
+        return;
+
+    NSMutableDictionary<NSString *, id> *changedData = [NSMutableDictionary dictionary];
```
auto

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm:
```diff
@@ -75,6 +79,21 @@
+    }
+
+    ASSERT_NOT_REACHED();
+    return local();
```
not sure what would be best to return here if this is reached.

xeenon replied:
This seems fine.

jeffmapple commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -305,6 +305,21 @@ id parseJSON(API::Data& json, JSONOptionSet options, NSError **error)
 }
 
+NSDictionary *keepObjectsWithKeys(NSDictionary *dictionary, NSArray *keys)
+{
```
I think you should implement this using `filterObjects()`.

jeffmapple commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -34,13 +34,18 @@
 #import <wtf/cocoa/VectorCocoa.h>
 
+static NSString * const newValueKey = @"newValue";
+static NSString * const oldValueKey = @"oldValue";
```
It looks like these keys are only used in `fireStorageChangedEventIfNeeded()`, if so you could move them inside that function scope.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -147,13 +141,20 @@
-        // FIXME: <https://webkit.org/b/267965> Fire onChanged event.
+        [storage deleteValuesForKeys:createNSArray(keys).get() completionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, storageType, oldValuesAndKeys = RetainPtr { oldValuesAndKeys }, completionHandler = WTFMove(completionHandler)](NSString *errorMessage) mutable {
+            if (errorMessage)
+                completionHandler(toErrorString(callingAPIName, nil, errorMessage));
```
Early return and no `else` would flow better and be more like other spots.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -167,13 +168,20 @@
-        // FIXME: <https://webkit.org/b/267965> Fire onChanged event.
+        [storage deleteDatabaseWithCompletionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, oldValuesAndKeys = RetainPtr { oldValuesAndKeys }, storageType, completionHandler = WTFMove(completionHandler)](NSString *errorMessage) mutable {
+            if (errorMessage)
+                completionHandler(toErrorString(callingAPIName, nil, errorMessage));
```
Ditto on early return.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -167,13 +168,20 @@
     auto storage = storageForType(storageType);
-    [storage deleteDatabaseWithCompletionHandler:makeBlockPtr([&, completionHandler = WTFMove(completionHandler)](NSString *errorMessage) mutable {
-        if (errorMessage)
+    [storage getValuesForKeys:@[] completionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, storageType, storage, completionHandler = WTFMove(completionHandler)](NSDictionary<NSString *, NSString *> *oldValuesAndKeys, NSString *errorMessage) mutable {
```
`@[ ]`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -191,6 +199,50 @@
+
+    auto *changedData = [NSMutableDictionary dictionary];
+
+    if (bool dataWasRemovedOrCleared = !newKeysAndValues) {
```
No need for the var.

```
if (!newKeysAndValues) {
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm:
```diff
@@ -85,6 +104,31 @@
+            namespaceObject.storage().onChanged().invokeListenersWithArgument(onChangedData);
+            namespaceObject.storage().storageAreaForType(storageType).onChanged().invokeListenersWithArgument(onChangedData);
+        });
+        break;
```
These could be combined to use `enumerateNamespaceObjects()` and pass the right world for the second optional argument, based on `contentWorldType`. A helper function not get the world based on `contentWorldType` would be good for this too.

kiaraarose commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -305,6 +305,19 @@ id parseJSON(API::Data& json, JSONOptionSet options, NSError **error)
+    return filterObjects(dictionary, ^bool(id key, id) {
+        return dictionary[key] && [keysSet containsObject:key];
+    });
+
```
remove extra newline

kiaraarose committed:
Fire onChanged events for Web Extension Storage API.
https://bugs.webkit.org/show_bug.cgi?id=267965
rdar://121474442

Reviewed by Timothy Hatcher.

This patch fires storage.onChanged and storageArea.onChanged events when new data is added or
old data is removed from storage. The onChanged data sent to the listener only contains data
for key value pairs where the value for that key has changed. If no data has changed, don't
fire the event.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::dictionaryWithKeys):
Helper method that returns a new dictionary based on the one provided, containing the specified keys.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
(WebKit::WebExtensionContext::storageSet):
(WebKit::WebExtensionContext::storageRemove):
(WebKit::WebExtensionContext::storageClear):
(WebKit::WebExtensionContext::fireStorageChangedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm:
(WebKit::WebExtensionAPIStorage::storageAreaForType):
(WebKit::WebExtensionContextProxy::dispatchStorageChangedEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/273942@main
